gadopt/equations.py

```diff
@@ -1,204 +1,177 @@
+                p = max(p)
 
-    Attributes:
-      terms (list[BaseTerm]): List of equation terms defined through inheritance from BaseTerm.
+            quad_degree = 2 * p + 1
 
-    Arguments:
-      test_space:  Firedrake function space of the test function
-      trial_space: Firedrake function space of the trial function
-      quad_degree:
-        Quadrature degree. Default value is `2p + 1`, where p the polynomial degree of the trial space
-      prefactor:   Constant prefactor multiplying all terms in the equation
+        self.mesh = self.trial_space.mesh()
+        self.n = fd.FacetNormal(self.mesh)
 
-    """
-    terms = []
+        measure_kwargs = {"domain": self.mesh, "degree": quad_degree}
+        self.dx = fd.dx(**measure_kwargs)
 
-    def __init__(
-        self,
-        test_space: firedrake.functionspaceimpl.WithGeometry,
-        trial_space: firedrake.functionspaceimpl.WithGeometry,
-        quad_degree: Optional[int] = None,
-        prefactor: Optional[Number | firedrake.Constant] = 1,
-        **kwargs
-    ):
-        self.test_space = test_space
-        self.trial_space = trial_space
-        self.mesh = trial_space.mesh()
-
-        p = trial_space.ufl_element().degree()
-        if isinstance(p, int):  # isotropic element
-            if quad_degree is None:
-                quad_degree = 2*p + 1
-        else:  # tensorproduct element
-            p_h, p_v = p
-            if quad_degree is None:
-                quad_degree = 2*max(p_h, p_v) + 1
-
-        if trial_space.extruded:
+        if self.trial_space.extruded:
             # Create surface measures that treat the bottom and top boundaries similarly
             # to lateral boundaries. This way, integration using the ds and dS measures
             # occurs over both horizontal and vertical boundaries, and we can also use
             # "bottom" and "top" as surface identifiers, for example, ds("top").
-            self.ds = CombinedSurfaceMeasure(self.mesh, quad_degree)
-            self.dS = (
-                firedrake.dS_v(domain=self.mesh, degree=quad_degree) +
-                firedrake.dS_h(domain=self.mesh, degree=quad_degree)
-            )
+            self.ds = CombinedSurfaceMeasure(**measure_kwargs)
+            self.dS = fd.dS_v(**measure_kwargs) + fd.dS_h(**measure_kwargs)
         else:
-            self.ds = firedrake.ds(domain=self.mesh, degree=quad_degree)
-            self.dS = firedrake.dS(domain=self.mesh, degree=quad_degree)
-
-        self.dx = firedrake.dx(domain=self.mesh, degree=quad_degree)
+            self.ds = fd.ds(**measure_kwargs)
+            self.dS = fd.dS(**measure_kwargs)
 
-        # General prefactor multiplying all terms in the equation
-        # N.b. setting this to a firedrake constant (i.e. prefactor = Constant(1)) breaks
-        # Drucker-Prager rheology test even though it is being multiplied by 1...
-        self.prefactor = prefactor
+    def mass(
```

Comment:
Any reason why you are calling this mass intead of mass_term?

Reply:
I think I wanted to reserve `_term` to functions defined in modules such as `momentum_equation.py` or `scalar_equation.py`. You can see below that I also just have `residual`. Is that an issue?

Reply:
ok, but isn't it still a mass_term. the reason you can put it in equations.py is because it is pretty general to any time dependent pde...

Reply:
I do not understand what `_term` brings here. Surely no one will confuse this with the physical mass. From the finite elements perspective, what we are implementing here is a mass matrix. Given that we define terms in other modules, `mass` seems appropriate to me here.